### DISCUSSION TITLE
Ignore empty patches when scrolling through voices

### DISCUSSION THREAD

**Discussion by donluca** [probonopd/MiniDexed#454] [Ideas]
Some sysex files have empty patches which you'll encounter while scrolling through the voices on Minidexed and, if normally they are not a big issue, when you have thousands of sysex files they really start to add up.

Can we put a check in place so that empty patches get skipped when scrolling through voices?

Thanks!

EDIT: now that I think about it... this could very well be filed as a bug as there's no good reason why empty patches should be ever shown, maybe I should have opened an Issue instead of a discussion.

**Comment by donluca** [comment]
This has been done, I've submitted a PR #466

**Comment by probonopd** [comment]
Do you know what a real DX7 does in this case?

**Reply by donluca** [reply]
I have no idea unfortunately, but this is more a QoL change I've implemented for myself rather than replicating a behavior of the original DX7.

I've made a PR because I think this would be useful to other people as well.

**Comment by BobanSpasic** [comment]
How do you define an empty patch? Is it the INIT VOICE patch? INIT VOICE is not empty.
Some banks are using other names for this same voice:
- 10x space
- 10x ~
- etc.

**Reply by donluca** [reply]
Yes, I've found banks which had lots of "INIT VOICE" patches and they were just "empty".

Bear in mind that "emtpy" doesn't mean that they do not make any sound, it means that it is initialized with "standard" parameters (no clue about what they are) and, as a consequence, they all sound the same.

So I took for granted that those were considered "empty" as in "initialized with standardized values" and filtered them out.

On Dexed, for example, if you create a new bank it will put "INIT VOICE" in all the patches and initialize them with those standard settings.

EDIT: look for example at DX7-DS55A-4.syx: it has only 4 patches and the rest are "INIT VOICE" which all sound the same and are basically "empty".

Please let me know any "alternative names" being used for those empty patches and I'll add them to the filter.

**Reply by diyelectromusic** [reply]
If they are filtered out then how can someone edit them to make them "non empty"?  If someone was designing a patch are we saying they can't start from MiniDexed?  Although I appreciate it isn't a particularly user friendly way to edit patches - but it is possible!

I guess the model is that all banks have 32 voices, but if we start filtering some out that sort of breaks the basic model of how the DX7 worked...?

**Reply by donluca** [reply]
Unfortunately this discussion is split across here and and #466 where I gave the solution for this issue:

"_From the wiki:_
> Open MiniDexed -> TG1 -> Bank
> Turn the knob right until you have selected a bank called NO NAME
> Open MiniDexed -> TG1 -> Voice
> Turn the knob right until you have selected a voice called INIT VOICE

_I can change my code to create an exception where if the bank name is "NO NAME", then it doesn't skip "INIT VOICE", so you can use that to create patches from scratch._"

Let's keep the discussion here if possible and keep the PR to discuss only technical issues @probonopd so we can avoid information fragmentation.

EDIT: actually, you "gave" me the solution already to this issue.
I can make this an option you can either enable or disable, although it should really be straight in the UI rather than in the configuration file, IMHO.

**Reply by probonopd** [reply]
Let's avoid too much complexity. I tend to say that if you load a bank with empty patches on your device, then you _want_ a bank with empty patches on your device. To me this discission is similar to saying "let's not show empty folders in the file manager". There are empty ones, we show empty ones. You don't like empty ones, you don't make empty ones. Easy. 

That some collections of patches on the Internet are low quality is nothing we should try to fix by making the instrument more complicated imho.

Or am I overlooking something?

**Reply by donluca** [reply]
The use case of this change I proposed was that once you have a substantial library of banks, even scrolling through just 100 banks starts to become quite tiresome (that's 3200 voices), especially if there are banks with lots of empty voices.

The fact that there are some banks with lots of empty voices, IMHO, has nothing to do with the quality of the bank: if the author made just 10 patches out of 32 it's not his fault that there are 22 empty patches you have to skip manually by turning the encoder over and over.

Besides, this is one of the changes I've made for myself which has made my experience in using MiniDexed with lots of banks way more enjoyable, so I wanted to see if people were interested in having this option.

If it is not deemed useful or can create issues then we can just scrap it and I'll just use it myself.

I think that having this as an option you can enable/disable could be a good solution as I'm sure I won't be the only one to load my MiniDexed with lots and lots of patches, many of which have unfortunately many empty banks to scroll through.

**Reply by probonopd** [reply]
> I can change my code to create an exception where if the bank name is "NO NAME", then it doesn't skip "INIT VOICE", so you can use that to create patches from scratch."

Can we make it so that it does not skip patches that contain the string `INIT` in their name (e.g.,: `INIT`, `INIT VOICE`, `NEW INIT`)? And let's not forget to document this behavior.

**Reply by donluca** [reply]
I'm a bit confused about what you're asking me, but that's probably my fault because I've been throwing around 200 different approaches to this, so it's better if I recap what the the PR #466 does as of now.

While scrolling voices (any voice of any bank), if the name of the voice is ```INIT VOICE``` then it gets skipped.
I made this change because there are banks which have 20+ empty voices and I don't want to scroll through all of them back and forth while browsing voices.

Now, it has been brought up an issue with this approach and it's that if anyone wants to create a patch from scratch on MiniDexed he won't be able to due to the aforementioned change.

There are two possible options to get around this:

1. Make this an option which can be enabled or disabled in the configuration file
2. MiniDexed offers a bank named ```NO NAME``` which is made explicitly for the user to make new voices from scratch, so I can add an exception so that if the user is in the ```NO NAME``` bank, then ```INIT VOICE``` do not get skipped.

If I understood correctly your request, you're asking me that voices which contain the string ```INIT``` in their name (which includes ```INIT VOICE```) do not get skipped, which means that... my change doesn't make any difference from the default behavior.

Please let me know if I understood correctly or if there's been a misunderstanding from my part.

**Reply by donluca** [reply]
Actually, I think I have found an incredibly elegant solution to this issue.

The real issue here is not whenever you encounter a single empty voice, but when there are a lot of empty voices one after the other.

All that's needed to have the possibility to create new voices is just **one** empty voice.

The solution is as follows:

- Check voice name
- Is the voice name ```INIT VOICE```?
- If yes, is it the first ```INIT VOICE``` found?
- If it is, then just show it
- If it is the second subsequent ```INIT VOICE``` then skip it and all the other subsequent ones, until a non-```INIT VOICE``` is found

This way you'll always have access to an empty voice so you can modify it/create your own, but remove any other subsequent empty voices.

Yes, I believe this is the way to go, hopefully I managed to explain it clear enough of how this is going to work.

@diyelectromusic @probonopd your feedback would be immensely appreciated.

**Reply by diyelectromusic** [reply]
That sounds good to me, but I have to confess as to not have been following this too closely (I'm not sure I understand the "NO NAME" bank think for example). I certainly like the idea of intelligently knowing when to skip rather than relying on config settings. I guess the hit from this is that every bank containing blank voices will always have one "INIT VOICE" shown, but that is better than a whole string of them I assume.

I haven't followed it through, but I guess things to check are the hard-coded bank/voice information that is used if there are no ROMS at all.   I think they are the equivalent to Brass 1 and the ROM1A voice set, so are probably ok, but just add "having no ROMS" as a test case just to be sure and that should be fine :)

Also I guess the interaction with the change I did to skip invalid/non-loaded banks.  Again I don't think there will be an issue, but that it is worth re-running any tests you've done on the build that includes that (if you haven't already).

Kevin

**Reply by probonopd** [reply]
To be honest, I am still not sure whether this is a good idea at all. How is one supposed to save a new voice in, e.g., place 17 of a bank that was previously empty? (In case we ever implement saving/writing back to banks, that is.) I'd always have assumed: Load the blank voice that is currently in place 17, edit it, and save it. It will be saved in place 17 of the bank. Now, when one cannot load the 17th voice in a bank just because there are other blank ones before it in the same bank, things start to get complicated.

**Reply by diyelectromusic** [reply]
I think this is starting to show two different sets of users here.  On the one hand we have the "load lots of presets and enjoy exploring them" type and on the other we have the "experiment with FM synth parameters for interesting sounds" type.  Both are valid, but I think benefit differently from this change being in or out... If I was stereotyping, I'd imagine the former are those who grew up with a DAW whereas the latter are those (like myself) who grew up with a synth with a limited number of presets and user memory slots :)

On that basis, a config option is probably the only way to resolve it!

Having just loaded up the new version of the curated set, I can't see how anyone could ever get through 2000+ banks of 32 voices to see what they sounded like. It took more than a minute to just scroll the banks with the rotary encoder with no time to read the names, let alone select any of the voices and hear what they sounded like!? :)

I guess one way to tame it is to provide a bank/voice list for something like MidiOx alongside the curated set so that they can be selected externally, but that is an awful lot of work and at that point you probably don't care if scrolling INIT voices is a thing or not!

Kevin

**Reply by probonopd** [reply]
> I think this is starting to show two different sets of users here. On the one hand we have the "load lots of presets and enjoy exploring them" type and on the other we have the "experiment with FM synth parameters for interesting sounds" type.

Yes, probably.

> On that basis, a config option is probably the only way to resolve it!

Possibly. Although in general I am not a fan of config options that result in different behavior (as opposed to e.g., setting some pins), because it is what makes systems "complicated" to explain.

Do the unwanted blank voices have the word "INIT" in their name? If not, that could be the key to distingusihing between (wanted) INIT/INIT VOICE patches and unwanted "empty" ones?

**Reply by diyelectromusic** [reply]
> Possibly. Although in general I am not a fan of config options that result in different behavior (as opposed to e.g., setting some pins), because it is what makes systems "complicated" to explain.

You should probably have rejected more of my submissions these past few months then :)

I do feel that MiniDexed is serving many different types of users these days, so if we can cater for different ways of using it, whilst more complicated (and practically impossible to test anymore) that probably isn't a bad thing as such.

<shrugs> I don't know - we're getting into development philosophy now really.  Flexibility vs "the one true way" :)

Kevin

**Reply by donluca** [reply]
@diyelectromusic makes great points and yes, I can definitely see 2 (maybe more?) kind of users who'd use MiniDexed in a different way.

Myself, personally, I prefer a more "useable" (strictly IMHO) interface to create new patches so I use MiniDexed to play all the various patches I find over the internet and, eventually, transfer to it patches I've made myself on Dexed or other softwares.

One thing I wanted to specify: "empty" patches are explicitly called ```INIT VOICE``` and nothing else.
There are no other names for that, it seems like it's a convention, so just filtering out exactly ```INIT VOICE``` will get rid of them.

One last thing: I've already done this modification for myself which I'm completely satisfied with and I've been discussing here how we could alter this in order to accomodate other use cases.

If you're against this change, even through an option, that's 100% fine by me and I won't invest any more time looking into alternatives and how to code them and we can consider this closed.

Otherwise I think that my solution to skip subsequent empty voices could be a nice workaround to keep all functionalities intact (plus, if you wish, an option to disable it, of course).

**Reply by diyelectromusic** [reply]
> One thing I wanted to specify: "empty" patches are explicitly called `INIT VOICE` and nothing else. There are no other names for that, it seems like it's a convention, so just filtering out exactly `INIT VOICE` will get rid of them.

That is in the MIDI implementation chart/etc for the DX7 as the initial value for voice parameters.

![DX Voice Init](https://user-images.githubusercontent.com/68612569/229630545-cafd55a8-2f9d-49d7-8b69-8c0e4adee836.png)

**Reply by donluca** [reply]
Thanks for confirming that!

Also, I've just realized thanks to #461 that MiniDexed doesn't actually write into the sysex files, but rather it stores the changes into performances.

That's quite the head scratcher! 😅 

Now I understand what @probonopd was saying in other discussions.

I'm still not 100% sure of how this all works though... looking at performances.ini, for example, I see that it still points to a bank + voice number, but it "overwrites it on the fly" with the parameters the user has input and saved into the performance? 🤔 

So it basically is "writing into the voices without actually writing the sysex file", hence we still need actual sysex files for users to create their custom voices which will then get saved into the performances.

Which means that the name you give to the bank and to the voices has absolutely no effect whatsoever, so we can safely create sysex files with voices initialized as described in your picture and instead of naming them ```INIT VOICE```, we can name them ```USER VOICE```.

The only issue I see is giving the user an easy way to make those "empty" sysex files... sure, we can provide them with one already, but then they'll have to manually duplicate it themselves, with the correct numbering.

I don't know honestly, but if we can already agree to change the "user voices" name to ```USER VOICE``` instead of ```INIT VOICE``` then my PR can already work as it is.

**Reply by diyelectromusic** [reply]
Well, seeing as they can't save voices to banks, it would be a very dedicated MiniDexed user to enter in all parameters for 32 voices on every power up.... so I think just providing one blank bank gives people the option to experiment with voices as part of the exercise if they want to.

(Update: Just to capture the whole conversion: more discussion about providing a "blank" ROM for users to edit here: https://github.com/probonopd/MiniDexed/pull/461#issuecomment-1494510648)

And even if they want 8 custom voices for a performance (one per TG), well that still leaves them 24 spare.

As I said in my other comment, my own opinion is that the save/load performances has changed (even broken) the mental model around banks/voices, but also as I've said before above, MiniDexed is no longer just a DX7 emulation and people are using 8 TGs as a single instrument, so its probably not a big deal.

Aside...

All "live" voices are loaded from SysEx files into a set of TVoiceBank structures accessed via CSysExFileLoader::m_pVoiceBank.  There are now 16384 of these "slots".  The SysEx file loader fills this in from the voices directory.  They are held in memory and not (yet) sent to the actual TG instances.

CMiniDexed::ProgramChange() will load a voice to the TG from this structure based on the bank and voice number using CSysExFileLoader::GetVoice() call to grab the data from the in-memory TVoiceBank and then Dexed::loadVoiceParameters() to write to the TG.  This can be called from either the UI or due to a MIDI ProgramChange message.

The performance system loads a single voice per TG if it is stored in the performance.ini file as text in a properties structure called CPerformanceConfig::m_nVoiceDataTxt (one per TG).  Then in MiniDexed the LoadPerformancePerameters() call uses VoiceDataFilled() to see if there is valid voice data in the performance property and if so decodes the text using GetVoiceDataFromTxt() and then loads those directly into the TG using the same Dexed::loadVoiceParameters() call that ProgramChange uses.

This all happens after CMiniDexed::loadPerformanceParameters() has already processed the Bank and Voice numbers, so it will have already called BankSelect and ProgramChange (just like the UI and MIDI can) but then just send over a new set of voice parameters regardless if it exists.

Confusingly there is also a CMiniDexed::loadVoiceParameters() call that calls Dexed::loadVoiceParameters() for the specified TG on receipt of a MIDI SysEx message containing voice data.  MiniDexed only supports single voice SysEx MIDI loading though, there is no code at present for bulk voice data.

So basically, the Dexed synth part knows nothing about all this - it just gets a block of voice data via the Dexed::loadVoiceParameters() call.

So is that all there is to know about voice parameters?  Not quite.  The UI also allows you to change voice parameters.  This happens via calls to CMiniDexed::Get(Set)VoiceParameter which work on the live loaded voice into Dexed using Dexed::get(set)VoiceElement.  These changes don't go anywhere near any of the above and only live "in" the TG itself.

As I say, it is all a bit confusing...

Kevin

**Reply by donluca** [reply]
>  The UI also allows you to change voice parameters. [...] These changes don't go anywhere near any of the above and only live "in" the TG itself.

Are you sure? I was under the impression that those gets saved in the performance...

Anyway, what's the plan then?
Just add a "blank" sysex file with voices named ```USER VOICE``` instead of ```INIT VOICE``` in getsysex.sh?
I can code it so that after it has finished downloading all the various banks from the internet it adds at the end the "blank"/"user" bank with empty voices.

**Reply by Banana71** [reply]
> Are you sure? I was under the impression that those gets saved in the performance...

It is true that: Changes to Voices via the UI are saved in the Performance.

**Reply by diyelectromusic** [reply]
They are saved in the performance.ini file by putting out of Dexed.  So the "live" voice data is saved as part of the performance, but not as a voice in its own right. I was only describing the loading process.

For saving, the data from the textual version: m_nVoiceDataTxt[] is saved in the performance file.  But before this, as part of CMiniDexed::DoSavePerformance, it calls Dexed::getVoiceData() to pull out the current state of the "live" voice data for the TG and converts it to "performance text" using SetVoiceDataToTxt.  So the bank/voice information is irrelevant really at this point. Even if you've not edited the data and are using a bank/voice "as is" it won't matter what the bank/voice numbers say on reloading it will load the voice data directly.  If you swap banks or even delete them, having the bank/voice numbers in the performance is meaningless after a save.

As I say, the performance model has confused the bank/voice model.

More on saving here: https://github.com/probonopd/MiniDexed/pull/461#issuecomment-1494799837

Kevin

**Reply by diyelectromusic** [reply]
> Anyway, what's the plan then? Just add a "blank" sysex file with voices named `USER VOICE` instead of `INIT VOICE` in getsysex.sh?

Given the confusing way saving is handled, I think this is by far the simplest option right now.  Maybe we could add saving in the future to a specific bank/voice, but even then having blank voice as the target of the save still feels like the simplest way to go in my view.

As I say, many synths from the time would have preset, cartridge, and "user" banks with only the "user" banks editable, so that feels ok to me.  The user just gets to decide which bank is the "user" bank and what to call it :)

Kevin

**Reply by diyelectromusic** [reply]
By the way, there is a risk here that someone might create a performance, save it, and then change the details in the ini file itself to change the bank/voice number and then wonder why it doesn't work.

Example: I create an 8-layered version of Brass 1 (bank 0/voice 1) and save it.  But then decide I'd like a string version to, so copy the performance.ini file to a new one and change the voice number to select strings.  However all the voice data in the copied file will still be that of Brass 1 so it will override the bank/voice selection.

The work around would be to remove the voice data in the copied file of course.  I don't know what to do about this other than to note it somewhere!

Kevin

**Reply by donluca** [reply]
> Given the confusing way saving is handled, I think this is by far the simplest option right now. Maybe we could add saving in the future to a specific bank/voice, but even then having blank voice as the target of the save still feels like the simplest way to go in my view.

Agreed.
I also think that people using Minidexed have become accustomed to the way it works, so I personally wouldn't touch performances or editing/saving/loading banks/voices.
The best thing we can do is document and then it will be up to the user to follow the documentation and find his way around doing whatever he wants to do with MiniDexed.

@probonopd if you give this solution your blessing, I'll modify the getsysex.sh script in order to add, at the end of the banks, one "user bank" filled with initialized voices named ```USER VOICE```.

This way you can safely merge PR #466 without issues as it will filter empty voices only in "pre-made" banks.

**Reply by probonopd** [reply]
> Also, I've just realized thanks to https://github.com/probonopd/MiniDexed/pull/461 that MiniDexed doesn't actually write into the sysex files

...yet. I still think it would be welcome if we could write changes back to sysex bank ("cartridge") files. It's just not implemented at the moment.

**Comment by Banana71** [comment]
I feel like we're talking past each other a bit.  There is no clear direction.  I think we should first document what we have and discuss what we want to achieve.  The miniDexed is a complex instrument and has many areas of application.  Whether you want to play a performance alone on the keyboard, you want to play music with a DAW or you want to scroll through thousands of sounds.  Each application has its justification and should be considered.

![PXL_20230404_112808006](https://user-images.githubusercontent.com/104296205/229779657-c8a1988a-5955-411e-bbeb-61a8b9795583.jpg)

Example of an overview
